starter_project/normalize_breton_lib.py

```diff
@@ -1,19 +1,133 @@
+MIXED_TRIGGERS = MUTATION_HANDLER.mixed_triggers
+NASAL_TRIGGERS = MUTATION_HANDLER.nasal_triggers
+LENITION_TRIGGERS = MUTATION_HANDLER.lenition_triggers
+ECLIPSIS_TRIGGERS = MUTATION_HANDLER.eclipsis_triggers
 
-def NormalizeBreton(breton_string):
-  return compose(breton_string, REMOVE_EXTRA_HYPHEN_IN_WORD).string()
+SOFT_MUTATION = MUTATION_HANDLER.soft_map
+HARD_MUTATION = MUTATION_HANDLER.hard_map
+SPIRANT_MUTATION = MUTATION_HANDLER.spirant_map
+MIXED_MUTATION = MUTATION_HANDLER.mixed_map
+NASAL_MUTATION = MUTATION_HANDLER.nasal_map
+LENITION_MUTATION = MUTATION_HANDLER.lenition_map
+ECLIPSIS_MUTATION = MUTATION_HANDLER.eclipsis_map
+
+SPACE = " "
+SIGMA_STAR = MUTATION_HANDLER.sigma_star
+
+PREPROCESS = MUTATION_HANDLER.preprocessing
+POSTPROCESS = MUTATION_HANDLER.postprocessing
+
+DO_SOFT_MUTATION = cdrewrite(
+    SOFT_MUTATION,
+    union(SPACE, "[BOS]") + SOFT_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_HARD_MUTATION = cdrewrite(
+    HARD_MUTATION,
+    union(SPACE, "[BOS]") + HARD_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_SPIRANT_MUTATION = cdrewrite(
+    SPIRANT_MUTATION,
+    union(SPACE, "[BOS]") + SPIRANT_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_MIXED_MUTATION = cdrewrite(
+    MIXED_MUTATION,
+    union(SPACE, "[BOS]") + MIXED_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_NASAL_MUTATION = cdrewrite(
+    NASAL_MUTATION,
+    union(SPACE, "[BOS]") + NASAL_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_LENITION = cdrewrite(
+    LENITION_MUTATION,
+    union(SPACE, "[BOS]") + LENITION_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_ECLIPSIS = cdrewrite(
+    ECLIPSIS_MUTATION,
+    union(SPACE, "[BOS]") + ECLIPSIS_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+
+MUTATION_DICT = {
+    "soft":
+        DO_SOFT_MUTATION,
+    "hard":
+        DO_HARD_MUTATION,
+    "spirant":
+        DO_SPIRANT_MUTATION,
+    "mixed":
+        DO_MIXED_MUTATION,
+    "nasal":
+        DO_NASAL_MUTATION,
+    "lenition":
+        DO_LENITION,
+    "eclipsis":
+        DO_ECLIPSIS,
+}
+
+DO_PREPROCESSING = cdrewrite(
+    PREPROCESS,
+    union(SPACE, "[BOS]"),
+    union(GRAPHEMES, SPACE),
+    SIGMA_STAR)
+
+DO_POSTPROCESSING = cdrewrite(
+    POSTPROCESS,
+    union(SPACE, "[BOS]"),
+    union(GRAPHEMES, SPACE),
+    SIGMA_STAR)
+
+
+def preprocess(string: str) -> str:
+    """Preprocess the string before normalization."""
+    return compose(string.strip().lower(), DO_PREPROCESSING).string()
+
+
+def postprocess(string: str) -> str:
+    """Postprocess the string after normalization."""
+    return compose(string, DO_POSTPROCESSING).string()
+
+
+def apply_mutation(string: str, mutation: str) -> str:
+    """Applies a single mutation to a string."""
+    applied_mutation = compose(string, MUTATION_DICT.get(mutation)).string()
+    return applied_mutation
+
+
+def normalize_breton(breton_string: str) -> str:
```

Comment:
Let's use FST composition here!

You could remove the definitions for the pre and post processing, keep the mutation definitions but just give them the "breton_string" argument, and then do:

return preprocess(breton_string) @ apply_soft_mutation @ apply_hard_mutation @ apply_spirant_mutation @ postprocess(breton_string)

For this to work though, IIUC you'll have to change your functions to output FSTs rather than strings, then you can output strings in the end.

Reply:
Changed `normalize_breton` function to this:
```
def normalize_breton(breton_string: str) -> str:
    """Applies Breton mutations."""
    return (breton_string.strip().lower() @
            DO_PREPROCESSING @
            DO_SOFT_MUTATION @
            DO_HARD_MUTATION @
            DO_SPIRANT_MUTATION @
            DO_POSTPROCESSING).string()
```
Also removed `apply_mutation`, `preprocess` and `postprocess` functions and `MUTATION_DICT` dictionary, since they are not used after the previous change makes it so we just call the cdrewrite rules directly.